tools/unicode_properties_parse/format_width_estimate_intervals.cpp

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// The following code generates data for `_Width_estimate_intervals` in <format>.
+
+#include <charconv>
+#include <format>
+#include <fstream>
+#include <iostream>
+#include <regex>
+#include <vector>
+
+void _verify(bool test, int line, const char* msg) {
+    if (!test) {
+        std::cerr << "Error at line " << line << ": " << msg << std::endl;
+        exit(EXIT_FAILURE);
+    }
+}
+#define verify(expr, msg) _verify((expr), __LINE__, (msg))
+static const char* impl_assertion_failed = "impl assertion failed";
+
+struct range_u {
+    uint32_t from, to;
+    range_u(uint32_t f, uint32_t t) : from(f), to(t) {}
+    range_u(uint32_t v) : from(v), to(v) {}
+};
+
+// A valid Unicode code point won't exceed `max_u`.
+const uint32_t max_u = 0x10'ffff;
+using table_u        = std::vector<bool>; // true: wide
+table_u make_table() {
+    return table_u(max_u + 1, false);
+}
+
+// For `_Width_estimate_intervals`.
+void print_intervals(const table_u& table) {
+    using namespace std;
+    cout << endl;
+    int c     = 0;
+    bool last = table[0];
+    for (uint32_t u = 0; u <= max_u; u++) {
+        if (table[u] != last) {
+            cout << "0x" << hex << uppercase << u << "u, ";
+            if (++c == 12) {
+                c = 0;
+                cout << endl;
+            }
+        }
+        last = table[u];
+    }
+    cout << endl;
+}
+
+void fill_range(table_u& table, const ::range_u rng, bool is_wide) {
+    const auto [from, to] = rng;
+    verify(from <= to && to <= max_u, impl_assertion_failed);
+    for (uint32_t u = from; u <= to; u++) {
+        table[u] = is_wide;
+    }
+}
+
+table_u get_table_cpp20() {

```

Comment:
Since the paper addresses https://cplusplus.github.io/LWG/issue3780 libc++ backported this paper to C++20. I'm not a MSVC STL maintainer so I don't know what MSVC STL wants to do.

Reply:
This PR is applying the changes to C++20 mode. I think the function names are improper, maybe we should change to `until_lwg_3780`/`since_lwg_3780`.

Although the "`until`" ("`cpp20`") version isn't used in the produce code...

Reply:
My mistake. Indeed the name is confusion. Especially since format was added in C++20.